drivers/CpuUid.cpp

```diff
@@ -0,0 +1,80 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "hal/cpu_uid_api.h"
+#include "platform/mbed_assert.h"
+#include "drivers/CpuUid.h"
+
+#if DEVICE_CPUUID
+
+namespace mbed {
+
+const char CpuUid::_hexChars[16] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+
+CpuUid::CpuUid() : _data(NULL)
+{
+    _size = cpu_uid_get_length();
+#ifdef MBED_ASSERT
+    MBED_ASSERT(_size > 0);
+#endif
+    if (0 < _size) {
+        _data = new uint8_t[_size];
+        cpu_uid_get_uid(_data);
+    }
+}
+
+CpuUid::~CpuUid()
+{
+    if (_data) {
+        delete _data;
+    }
+}
+
+CpuUid::operator std::string()
+{
```

Comment:
In many targets, the UID is already ASCII. The conversion to string should be target-defined.

Reply:
the three targets I have here (Silicon Labs EFM32, Nordic nRF52, STM32 Nucleo boards) all measure their UIDs in bits, like "96 bits Unique ID" for STM32 CPUs. And the values I get don't look like ASCII at all.
Which targets do you mean?

Reply:
For example: ST RM0090, Section 39.1:
* UID(31:0): X and Y coordinates on the wafer. 
* UID(39:32): WAF_NUM[7:0] Wafer number (ASCII encoded).
* UID(63:40): LOT_NUM[23:0] Lot number (ASCII encoded).
* UID(95:64): LOT_NUM[55:24] Lot number (ASCII encoded).

Reply:
That's right, but in each case the length is still known at compile time.

Reply:
Not necessarily, if target developers decide to do it "their own way" ... for example one vendor could decide to strip leading zeros, another could leave it. And then the string length is not known at compile time because dependant of the uid itself...

Reply:
This is where defining the behaviour is helpful. Define that the vendor must provide the max-size of the UID printable string in a macro. This makes everyone's life easier.

Reply:
yep, good idea

Reply:
but ...: if I would add the get_c_str() function, you mentioned above, to the HAL, then there is no need for the user to define its own buffer .... each target implementation shall define it's own static buffer, and they know what the max size should be. It gets filled, the first time someone calls the get_c_str() function, and gets only returned as const char*, so noone outside HAL implementation can mess with it.
Then there is no need to return a length ... if someone wants to know the length of the string he/she can do a strlen(uidstring);

Reply:
Even better!